Reference Guide for Cookieplone make commands (#1934)
plone/documentation: docs/reference-guide/cookieplone-make-commands.md

---
myst:
  html_meta:
    "description": "Cookieplone make commands"
    "property=og:description": "Cookieplone make commands"
    "property=og:title": "Cookieplone make commands"
    "keywords": "Cookieplone, make, commands"
---

# Cookieplone `make` commands

This reference guide describes the function and purpose of all the `make` commands found in a {term}`Cookieplone` project.
It's organized according to its components.
-   Frontend
-   Backend
-   Documentation
-   Environment
-   Quality assurance (QA)
-   Internationalization (i18n)
-   Testing
-   Container images
-   Local stack
-   Acceptance

## Frontend

When you issue a `make` command at the root of your project, you call the file {file}`Makefile` also at the root.
In turn, it invokes commands in the file {file}`frontend/Makefile`.
You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`frontend-install`
:   Invokes the target `install` in `frontend/Makefile`.

>     Uses pnpm to run the mrs-developer tool with the default settings.
>     It then installs dependencies using pnpm.
>     After that, it runs `build-deps` to build dependency packages.  

Reviewer: Not quite. See https://pnpm.io/cli/dlx and https://github.com/collective/mrs-developer. I had to look them up myself.

```make
	pnpm dlx mrs-developer missdev --no-config --fetch-https
	pnpm i
	make build-deps
```

```suggestion
    Uses `pnpm dlx` to run [mrs-developer](https://github.com/collective/mrs-developer) to retrieve dependencies in the file {file}`frontend/mrs.developer.json`.
    It then installs the dependencies.
    Finally, it builds the dependency packages `@plone/registry` and `@plone/components` from source.
```

You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`install`
:   Invokes the target `backend-install` and `make frontend-install` in `backend/Makefile` and `frontend/Makefile` respectively.
    This installs the add-ons in the development environment for both backend and frontend.

`clean`
:   Invokes the target `clean` in both `backend/Makefile` and `frontend/Makefile`.
    This command cleans both backend (removing virtual environments, cached files, and instance data) and frontend (removing core files and node_modules) environments.

## Quality assurance (QA)

When you issue a `make` command at the root of your project, you call the file {file}`Makefile` also at the root.
In turn, it invokes commands in the files {file}`backend/Makefile` and {file}`frontend/Makefile`.
You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`format`
:   Invokes the target `format` in both `backend/Makefile` and `frontend/Makefile`. 

>     This formats the code base according to Plone standards

Reviewer: You manually updated the code from my suggestion, so you missed the period. Do you know how to use GitHub to add suggestions to a batch, then commit? See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request.

```suggestion
    This formats the code base according to Plone standards.
```

Author: Thanks for the tip, I know about this but I usually just open my IDE do the changes there. I will keep it in mind.

---
myst:
  html_meta:
    "description": "Cookieplone make commands"
    "property=og:description": "Cookieplone make commands"
    "property=og:title": "Cookieplone make commands"
    "keywords": "Cookieplone, make, commands"
---

# Cookieplone `make` commands

> This guide explains all the `make` commands found in a cookieplone project.

Reviewer: ```suggestion
This reference guide describes the function and purpose of all the `make` commands found in a {term}`Cookieplone` project.
It's organized according to its components.

-   Frontend
-   Backend
-   Documentation
-   Environment
-   Quality assurance (QA)
-   Internationalization (i18n)
-   Testing
-   Container images
-   Local stack
-   Acceptance

```

## Frontend

> The Makefile at the root of your project invokes commands in frontend/Makefile. Here are excerpts from frontend/Makefile to show details of the frontend make commands.

Reviewer: ```suggestion
When you issue a `make` command at the root of your project, you call the file {file}`Makefile` also at the root.
In turn, it invokes commands in the file {file}`frontend/Makefile`.
You can refer to these files for implementation details.

You can run the following make targets by using the command structure of `make <TARGET>`.
```

> `frontend-install`
> 
> The command `make frontend-install`:
> 
> -   Invokes the target `make install` in `frontend/Makefile`.
>     -   This installs the add-ons in the development environment.
> 

Reviewer: Let's use a definition list for each target. It's cleaner and easier to read. Here's one example. I'll leave the rest for you to perform.

```suggestion
`frontend-install`
:   Invokes the target `make install` in `frontend/Makefile`.
    This installs the add-ons in the development environment.

```

`frontend-build`

The command `make frontend-build`:

-   Invokes the target `make build` in `frontend/Makefile`.
    -   This creates a production bundle for distribution of the project with the add-on.

`frontend-start`

The command `make frontend-start`:

-   Invokes the target `make start` in `frontend/Makefile`.
    -   This starts Volto, allowing reloading of the add-on during development.

`frontend-test`

The command `make frontend-test`:

-   Invokes the target `make test` in `frontend/Makefile`.
    -   This runs unit tests.

> ## Backend

Reviewer: Follow the example of Frontend.

# Cookieplone `make` commands

This reference guide describes the function and purpose of all the `make` commands found in a {term}`Cookieplone` project.
It's organized according to its components.
-   Frontend
-   Backend
-   Documentation
-   Environment
-   Quality assurance (QA)
-   Internationalization (i18n)
-   Testing
-   Container images
-   Local stack
-   Acceptance

## Frontend

When you issue a `make` command at the root of your project, you call the file {file}`Makefile` also at the root.
In turn, it invokes commands in the file {file}`frontend/Makefile`.
You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`frontend-install`

> :   Invokes the target `make install` in `frontend/Makefile`.

Reviewer: For this and all targets, please remove `make `. I must have been half asleep when I suggested that. `make install` is a command. `install` is a make target. See line 30 and https://makefiletutorial.com/

```suggestion
:   Invokes the target `install` in `frontend/Makefile`.
```

>     This installs the add-ons in the development environment.

Reviewer: I think it does much more than that. Take a look at the chain of commands in the Makefiles and console output of a freshly generated project.

Author: I know it does, but I wanted to keep the description short. Do you think I should add one or two lines, or make it more detailed?

Reviewer: There's a balance between "knowing how sausage is made" and "not enough information". I think it's important to mention some more details, such as what is in a development environment.

`frontend-build`
:   Invokes the target `make build` in `frontend/Makefile`.
    This creates a production bundle for distribution of the project with the add-on.

`frontend-start`
:   Invokes the target `make start` in `frontend/Makefile`.
    This starts Volto, allowing reloading of the add-on during development.

`frontend-test`
:   Invokes the target `make test` in `frontend/Makefile`.
    This runs unit tests.

## Backend

When you issue a `make` command at the root of your project, you call the file {file}`Makefile` also at the root.
In turn, it invokes commands in the file {file}`backend/Makefile`.
You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`backend-install`
:   Invokes the target `make install` in `backend/Makefile`.
    This creates a `Python` virtual environment if one does not exist.
    It then installs Plone and its dependencies in that virtual environment.

>     After installation, it runs `make backend-create-site` to initialize a new Plone site.

Reviewer: ```suggestion
    After installation, it runs `make backend-create-site` to initialize a new Plone site with default content.
```

`backend-build`
:   Invokes the target `make install` in `backend/Makefile`.

>     This creates a production bundle for distribution of the project with the add-on.

Reviewer: I don't think that's correct. I don't see it making a production bundle, but it just builds Plone. Maybe I missed something?

Author: You are right, it just builds Plone. I'm confused about what to write here. Can you help me?

Reviewer: ```suggestion
    This creates a `Python` virtual environment if one does not exist.
    It then installs Plone and its dependencies in that virtual environment.
    This is useful when you need to reload changes from your backend add-on, and don't need to recreate a full Plone site.
```

`backend-create-site`
:   Invokes the target `make create-site` in `backend/Makefile`.

>     This first ensures the virtual env exists.

Reviewer: ```suggestion
    This first ensures the virtual environment exists.
```

    Creates a new Plone site with default content.
  
`backend-update-example-content`
:   Invokes the target `make update-example-content` in `backend/Makefile`.

>     This first ensures the virtual env exists.
>     Checks if example-content exists. If yes, deletes all files inside to ensure a fresh export.
>     Then exports the Plone content.

Reviewer: ```suggestion
    This first ensures the virtual environment exists.
    Next, it removes all content from the destination directory, if any content exists.
    Finally, it exports the Plone site content.
```

`backend-start`
:   Invokes the target `make start` in `backend/Makefile`.

>     This starts a Plone instance on localhost:8080.

Reviewer: ```suggestion
    This starts a Plone instance on `localhost:8080`.
```

You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`install`
:   Invokes the target `make backend-install` and `make frontend-install` in `backend/Makefile` and `frontend/Makefile` respectively.
    This installs the add-ons in the development environment for both backend and frontend.

`clean`
:   Invokes the target `make clean` in both `backend/Makefile` and `frontend/Makefile`.
    This command cleans both backend (removing virtual environments, cached files, and instance data) and frontend (removing core files and node_modules) environments.

## Quality assurance (QA)

When you issue a `make` command at the root of your project, you call the file {file}`Makefile` also at the root.
In turn, it invokes commands in the files {file}`backend/Makefile` and {file}`frontend/Makefile`.
You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`format`
:   Invokes the target `make format` in both `backend/Makefile` and `frontend/Makefile`. 

>     This formats the codebase according to Plone standards

Reviewer: ```suggestion
    This formats the code base according to Plone standards.
```

`lint`
:   Invokes the target `make lint` in both `backend/Makefile` and `frontend/Makefile`. 

>     Checks for problems (linting, formatting, style issues) but does not auto-fix them, it only reports errors.

Reviewer: ```suggestion
    Checks for problems—such as linting, formatting, and style issues—but does not auto-fix them.
    It only reports errors.
```

`check`
:   Runs both `format` and `lint` in sequence.

## Internationalization (i18n)

When you issue a `make` command at the root of your project, you call the file {file}`Makefile` also at the root.
In turn, it invokes commands in the files {file}`backend/Makefile` and {file}`frontend/Makefile`.
You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`i18n`
:   Invokes the target `make i18n` in both `backend/Makefile` and `frontend/Makefile`. 

>     It is used to update locales in your project.

Reviewer: ```suggestion
    It updates translations in your project.
```

## Testing

When you issue a `make` command at the root of your project, you call the file {file}`Makefile` also at the root.
In turn, it invokes commands in the files {file}`backend/Makefile` and {file}`frontend/Makefile`.
You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`test`
:   Invokes the target `make backend-test` and `make frontend-test` in `backend/Makefile` and `frontend/Makefile` respectively. 
    This runs unit tests in the project.

## Container images

When you issue a `make` command at the root of your project, you call the file {file}`Makefile` also at the root.
In turn, it invokes commands in the files {file}`backend/Makefile` and {file}`frontend/Makefile`.
You can refer to these files for implementation details.
You can run the following make targets by using the command structure of `make <TARGET>`.

`build-images`
:   Invokes the target `make build-image` in both `backend/Makefile` and `frontend/Makefile`. 

>     This build Docker Images for both backend and frontend.

Reviewer: ```suggestion
    This builds Docker images for both backend and frontend.
```
